Cython/Compiler/ExprNodes.py

```diff
@@ -12249,9 +12269,36 @@ def analyse_c_operation(self, env):
                             (self.operand1.type, self.operand2.type))
 
     def compute_c_result_type(self, type1, type2):
-        c_result_type = super(PowNode, self).compute_c_result_type(type1, type2)
-        if isinstance(self.operand2.constant_result, _py_int_types) and self.operand2.constant_result < 0:
-            c_result_type = PyrexTypes.widest_numeric_type(c_result_type, PyrexTypes.c_double_type)
+        c_result_type = None
+        op1_is_definitely_positive = (
+            self.operand1.constant_result is not not_a_constant
+            and self.operand1.constant_result >= 0
+        ) or (
+            type1.is_int and type1.signed == 0  # definitely unsigned
+        )
+        type2_is_int = type2.is_int or (
+            self.operand2.constant_result is not not_a_constant and
+            int(self.operand2.constant_result) == self.operand2.constant_result
+        )
+        # if type2 is an integer than we can't end up going from real to complex
+        if self.cpow or op1_is_definitely_positive or type2_is_int:
+            c_result_type = super(PowNode, self).compute_c_result_type(type1, type2)
+            if self.operand2.constant_result is not_a_constant:
+                needs_widening = not self.cpow and type2.is_int and type2.signed
+                if needs_widening:
+                    self.cpow_false_changed_result_type = True
```

Comment:
At first (& second) glance, that inner `needs_widening` looks unreachable. On third thought I guess it isn't due to the various `or`-combinations in the initial if-condition.

Perhaps it would be easier to reason about (even if perhaps the code gets slightly longer) to separate the treatment of a conscious choice of nondefault `@cpow` from the could-the-result-be-complex inference.

Reply:
I've made this change